Mathlib/SetTheory/Lists.lean

```diff
@@ -164,55 +168,58 @@ theorem mem_cons {a y l} : a ∈ @cons α y l ↔ a ~ y ∨ a ∈ l := by
 
 theorem cons_subset {a} {l₁ l₂ : Lists' α true} : Lists'.cons a l₁ ⊆ l₂ ↔ a ∈ l₂ ∧ l₁ ⊆ l₂ :=
   by
-  refine' ⟨fun h => _, fun ⟨⟨a', m, e⟩, s⟩ => subset.cons e m s⟩
+  refine' ⟨fun h => _, fun ⟨⟨a', m, e⟩, s⟩ => Subset.cons e m s⟩
   generalize h' : Lists'.cons a l₁ = l₁' at h
   cases' h with l a' a'' l l' e m s;
   · cases a
     cases h'
   cases a; cases a'; cases h'; exact ⟨⟨_, m, e⟩, s⟩
 #align lists'.cons_subset Lists'.cons_subset
 
-theorem of_list_subset {l₁ l₂ : List (Lists α)} (h : l₁ ⊆ l₂) :
+theorem ofList_subset {l₁ l₂ : List (Lists α)} (h : l₁ ⊆ l₂) :
     Lists'.ofList l₁ ⊆ Lists'.ofList l₂ := by
-  induction l₁; · exact subset.nil
-  refine' subset.cons (Lists.Equiv.refl _) _ (l₁_ih (List.subset_of_cons_subset h))
+  induction' l₁ with _ _ l₁_ih; · exact Subset.nil
+  refine' Subset.cons (Lists.Equiv.refl _) _ (l₁_ih (List.subset_of_cons_subset h))
   simp at h; simp [h]
-#align lists'.of_list_subset Lists'.of_list_subset
+#align lists'.of_list_subset Lists'.ofList_subset
 
 @[refl]
 theorem Subset.refl {l : Lists' α true} : l ⊆ l := by
-  rw [← Lists'.of_to_list l] <;> exact of_list_subset (List.Subset.refl _)
+  rw [← Lists'.of_toList l]; exact ofList_subset (List.Subset.refl _)
 #align lists'.subset.refl Lists'.Subset.refl
 
 theorem subset_nil {l : Lists' α true} : l ⊆ Lists'.nil → l = Lists'.nil :=
   by
-  rw [← of_to_list l]
-  induction to_list l <;> intro h; · rfl
+  rw [← of_toList l]
+  induction toList l <;> intro h; · rfl
   rcases cons_subset.1 h with ⟨⟨_, ⟨⟩, _⟩, _⟩
 #align lists'.subset_nil Lists'.subset_nil
 
-theorem mem_of_subset' {a} {l₁ l₂ : Lists' α true} (s : l₁ ⊆ l₂) (h : a ∈ l₁.toList) : a ∈ l₂ :=
-  by
-  induction' s with _ a a' l l' e m s IH; · cases h
-  simp at h; rcases h with (rfl | h)
-  exacts[⟨_, m, e⟩, IH h]
+theorem mem_of_subset' {a} {l₁ l₂ : Lists' α true} (s : l₁ ⊆ l₂) (h : a ∈ l₁.toList) : a ∈ l₂ := by
+  -- Porting note: Previous code was:
+  -- induction' s with _ a a' l l' e m s IH; · cases h
+  -- simp at h; rcases h with (rfl | h)
+  -- exacts[⟨_, m, e⟩, IH h]
+  --
+  -- 'induction' tactic does not support mutually inductive types
+  admit
 #align lists'.mem_of_subset' Lists'.mem_of_subset'
```

Comment:
I gave it a shot using explicit recursion, using this:
```
theorem mem_of_subset' {a} : ∀ {l₁ l₂ : Lists' α true} (s : l₁ ⊆ l₂) (h : a ∈ l₁.toList), a ∈ l₂
  | .nil, l₂, Lists'.Subset.nil, h => by cases h
  | .(cons a l), .(l'), @Lists'.Subset.cons _ a a' l l' e m s, h => by
    simp at h
    rcases h with (rfl | h)
    case inl => exact ⟨_, m, e⟩
    case inr => exact mem_of_subset' s h
```
but Lean doesn’t find the recursion argument:
```
fail to show termination for
  
with errors
argument #3 was not used because its type is an inductive family and indices are not variables
  

argument #4 was not used because its type is an inductive family and indices are not variables
  

argument #6 was not used because its type is an inductive family and indices are not variables
  

structural recursion cannot be used

failed to prove termination, use `termination_by` to specify a well-founded relation
```

🤷🏻 was worth a try
```

Reply:
Found a solution, explicit recursion on `l₁` works.